Converting fmt.Errorf to new error code package format

mikebrow opened the pull request:
Convert "utils" static error strings to the new error code package format

Signed-off-by: Mike Brown brownwm@us.ibm.com

mikebrow committed:
Converting fmt.Errorf to new error code package format

Signed-off-by: Mike Brown <brownwm@us.ibm.com>

cpuguy83 commented:
Looks good, but I think the HTTP status codes should probably be 500 instead of 409 for these.

mikebrow commented:
NP... changed to 500 Internal Server Error.   Was using conflict with the expectation that the issue was one that could be recognized and corrected by the user.

calavera commented:
LGTM

LK4D4 commented on errors/utils.go:
```diff
@@ -0,0 +1,84 @@
+	ErrorCodeUsingGitCheckout = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "USINGGITCHECKOUT",
+		Message:        "Error trying to use git: %s (%s)",
+		Description:    "The specified error was thrown while using git checkout",
```
Why some description starts with `The specified error was thrown while` and some not?

mikebrow replied:
The specified error refers to the variable portion %s of the error message.  If the %s was for an image name it would say the specified image name...  The description is further describing the output error message.

LK4D4 commented:
I personally hate this, I don't know why we doing this.
Now each person who want to add any line to code will be forced to create that stuff in errcode.
There is no point in almost 99% percents of all those errors, it hurts reading, it hurts writing.

duglin commented:
@LK4D4 there are a couple of reasons behind this:
- move the static error strings in the code into a central place so that each has a well defined (comparable) ID
- this allows us to remove code like: https://github.com/docker/docker/blob/master/api/server/server.go#L233  . Doing strcmp instead of just simple ID compares is much safer since we can now change the error message text w/o needing to adhere to some funky syntax rules to ensure the strcmp work. There are several places in the code (not just that link) where we do strcmp and its risky/scary.
- this allows us to associate a non-500 http status code with any error condition that doesn't want to return 500
- this allows us to associate more descriptive text with each error condition. While we're not doing it today, I'm hoping that when we're done we can use these to generate docs for the errors we generate.  Ideally, we'd also return the ID for each error back to the client too, not just the text so they can use that to easily find it in the docs.  But this is all still to-be-designed-discussed - but the option is there.
- this lays the foundation for translation. By moving all of the error text into a look-up type of structure, we can move to a real i18n package at some point which would allow us to translate our errors into other languages with no impact on our main code, things should be isolate to the error code package stuff.

duglin commented:
@LK4D4 edited previous comment - one heck of a typo in there  :-)

LK4D4 commented:
@duglin I understand. But I'm failing to observer how utils package related to http status codes or comparing something in api package.
My understanding that such errors should be returned only from functions which directly called from api/server or we will end with project to which nobody wants to contribute, because it uses its own language.

duglin commented:
If you're saying that we should only use the errorCode package for errors that are eventually returned back to the CLI, then that is probably true from a purist perspective. Which is why we're not touching the 'pkg' dir since that's meant to be reused by things outside of Docker.

But, its not clear to me at any given `return fmt.Errorf("...")` line whether that will stop in its caller or will bubble back up to api/server. From what I've seen so far, it seems like in most cases, upon getting an error from a func, we just return it back up the stack.  There are some cases where we wrap them in another error, but there really aren't that many of those in comparison.  I suppose we could go through the exercise of examining all callers of a function to see if they bubble vs stop the 'error' and only use errorCode if its never bubbled, but that seems error-prone (pardon the pun), since any new func or modification of a calling func might one day bubble the error w/o realizing they need to examine ALL of the funcs they call to see if they use fmt.Errorf() instead of errorCode.  So, for consistency, and safety, it seemed right to just convert all engine error strings into errorCodes.  Now, docker devs should just know that each new error they generate needs a corresponding errorCode - a little more work, but adding the descriptive text is probably a good thing to do anyway - not to mention this might force people to reuse more errors instead of creating near duplicates.

LK4D4 commented:
@duglin it doesn't seem right for me to convert all errors. I think `errorCode` should be used only in `api/server`. It's fair balance as for me.
So, I'm NOT LGTM about all PRs which do this with functions which isn't called from `api/server`. We should discuss this with other maintainers.

duglin commented:
@LK4D4 if everything below api/server is just a string, then we have what we have today - strings and strcmp to see what it is. I'm not sure how else to do this w/o converting each string to a well defined type/ID.

LK4D4 commented:
I'm sure that converting all errors will hurt project, it will be ugly and alien to other golang programmers. We need to find balance.

cpuguy83 commented:
@LK4D4 +1 agree.
I think return real typed errors as the stdlib does is the way to go for interactions between daemon and it's subsystems.
Then errcode can be used between daemon and the API.

duglin commented:
I actually think it'll help in the long run because:
- it will do/enable everything I mentioned above
- it forces us to be explicit about what errors we return - which is all part of having a well defined API. Right now we're kind of ad-hoc with what we expose to the CLI and, eventually, people will demand stability and well documented APIs - and the exact list of possible errors (with descriptions) IMO is all part of that.

duglin commented:
@cpuguy83 I had a similar irc chat with stevvooe  about that and that is something we can move towards. But we need to remember that right now the engine, for better or worse, is one big blob w.r.t. errors. Meaning, whatever is returned by an engine-owned (sub)component is bubbled all the way back up to the http layer.  We can look at changing this but that's a pretty big redesign of the error handling within the engine and we'd need to then decide which of the sub-components we **don't** want to enable the stuff I listed above.  It might be hard to explain why we're not providing descriptions or non-english text for errors that Docker itself is generating. For component we don't own, I think its fair to say its out of our control, but not code we develop.

LK4D4 commented:
@mikebrow @duglin can we close this?

mikebrow commented:
Yes please.. I'll fix and resubmit after we have the final plan/design.
